sympy/polys/series/base.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+from typing import Protocol, Sequence, TypeVar
+from sympy.core.symbol import Symbol
+from sympy.polys.domains import Domain
+from sympy.polys.domains.domain import Er
+
+
+_x: Symbol = Symbol("x")
+TSeries = TypeVar("TSeries")
+
+
+def series_pprint(series: list[Er], prec: int | None) -> str:
+    """Convert a list of coefficients into a string representation of a power series."""
+    terms = []
+
+    if prec == 0:
+        return f"O({_x}**{prec})"
+
+    for i, coeff in enumerate(series):
+        if coeff == 0:
+            continue
+        if coeff == 1 and i != 0:
+            term = f"{_x}" if i == 1 else f"{_x}**{i}"
+        elif coeff == -1 and i != 0:
+            term = f"-{_x}" if i == 1 else f"-{_x}**{i}"
+        else:
+            if i == 0:
+                term = f"{coeff}"
+            elif i == 1:
+                term = f"{coeff}*{_x}"
+            else:
+                term = f"{coeff}*{_x}**{i}"
+        terms.append(term)
+
+    if not terms:
+        if prec is not None:
+            return f"0 + O({_x}**{prec})"
+        else:
+            return "0"
+
+    poly = " + ".join(term for term in terms).replace("+ -", "- ")
+    if prec is not None:
+        return f"{poly} + O({_x}**{prec})"
+    return poly
+
+
+class PowerSeriesRingProto(Protocol[TSeries, Er]):
+    """A protocol for a power series ring."""
+
+    def __init__(self, prec: int = 6, /) -> None:
+        """Initialize a power series ring over ZZ."""
+        ...
+
+    def __repr__(self, /) -> str:
+        """Return string representation of the ring."""
+        ...
+
+    def __call__(
+        self, coeffs: Sequence, prec: int | None = None
```

Comment:
We can leave this to a followup PR but using `Sequence` brings back `Any`.

This should be `Sequence[Er | int]` but then there would need to be a subclass Protocol for QQ that extends it to something like `Sequence[Er | int | tuple[int, int]]`.